src/nodes/dir.rs

```diff
@@ -181,16 +279,26 @@ impl Node for Dir {
         for entry in entries {
             let entry = entry?;
             let name = entry.file_name();
-            let fs_attr = entry.metadata()?;
+
+            if let Some(dirent) = state.children.get(&name) {
+                if dirent.explicit_mapping {
+                    continue;
+                }
+            }
 
             let path = state.underlying_path.as_ref().unwrap().join(&name);
+            let fs_attr = entry.metadata()?;
+            let fs_type = conv::filetype_fs_to_fuse(&path, fs_attr.file_type());
             let child = cache.get_or_create(ids, &path, &fs_attr, self.writable);
 
-            reply.add(child.inode(), pos,
-                conv::filetype_fs_to_fuse(&path, fs_attr.file_type()), &name);
+            reply.add(child.inode(), pos, fs_type, &name);
             // Do the insertion into state.children after calling reply.add() to be able to move
             // the name into the key without having to copy it again.
-            state.children.insert(name, child.clone());
+            let dirent = Dirent {
+                node: child.clone(),
+                explicit_mapping: false,
+            };
+            state.children.insert(name, dirent);
```

Comment:
Where do we remove deleted entries from `state.children`?

Reply:
Good question. Nowhere at the moment. The Go variant doesn't do this either so I'd rather not change the behavior at this point to keep them consistent. Added a TODO.

The reason this doesn't seem to be a problem is two-fold: first, during readdir, we never use these stale entries because they are not marked as `explicit_mapping` so they never leak; and, second, even if we encounter any stale entry at lookup time we'll hit an `ENOENT` when doing the stat there, and lookup is a prerequisite for any other operation on the entry.

So... I don't know if this problem is even observable from the outside, but I may be missing some corner case. Would be nice to fix.